### PR TITLE
Size 1 Dimensions in Layer Norm Fusion

### PR DESCRIPTION
**Pull request by philipphack**
Enables the fusion of layer norm patterns with degenerate input dimensions of size 1 and with additional optional type conversions or reshapes adding or removing degenerate dimensions.

### CODE REVIEW

**Comment by philipphack** [comment]
CC @reedwm.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 1]
```diff
@@ -930,23 +959,32 @@ class CudnnNormRewriterVisitor : public DfsHloRewriteVisitor {
...
-      std::vector<int64_t> non_norm_dims;
-      for (int64_t x_dim = 0; x_dim < x.Instr()->shape().rank(); ++x_dim) {
+      std::vector<int64_t> non_norm_dims, non_norm_dims_adjusted;
+      for (int64_t x_dim = 0, degen_dims = 0; x_dim < x.Instr()->shape().rank();
+           ++x_dim) {
         if (std::find(norm_dims.begin(), norm_dims.end(), x_dim) ==
             norm_dims.end()) {
           non_norm_dims.emplace_back(x_dim);
+          if (x.Instr()->shape().dimensions(x_dim) != 1) {
+            non_norm_dims_adjusted.emplace_back(x_dim - degen_dims);
```
Comment (reedwm) [inline xla/service/gpu/cudnn_norm_rewriter.cc:983]: The logic for non_norm_dims_adjusted is a bit hard to follow. Can you create an overload of `AdjustedDimensions` that takes a `Shape` and `Span<int>` of dimensions (with a clear comment describing what it does), have the version taking an instruction call this alternative version, and also call the alternative version here on `non_norm_dims` to get `non_norm_dims_adjusted`?

File: xla/service/gpu/cudnn_norm_rewriter_test.cc [hunk 2]
```diff
@@ -1088,6 +1375,125 @@ TEST_F(CudnnNormRewriterTest, LayerNormTrainBackward4D12) {
   TestNorm(hlo_text, optimized_hlo);
 }
 
+// Dbias not expected to be fused into norm gradient.
```
Comment (reedwm) [inline xla/service/gpu/cudnn_norm_rewriter_test.cc:1378]: Comment this is because dbias is not fused when a reduction dimension is degenerate.

Also, why is Dbias not fused in this case?

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 3]
```diff
@@ -568,12 +589,13 @@ auto Expectation(UniqueHloInstruction* expectation, Pattern pattern) {
 template <typename Pattern>
 auto Expectation(UniqueHloInstruction* expectation, HloInstruction** reduce,
                  Pattern pattern) {
-  auto shared_subpattern = MultiplyAnyOrder(m::Broadcast(m::ConstantScalar()),
-                                            AddReduce(reduce, pattern))
-                               .WithPredicate([](const HloInstruction* instr) {
-                                 return CalculatesExpectation(instr);
-                               })
-                               .WithPredicate(expectation->capture_or_verify);
+  auto shared_subpattern = OptionalSupportedTransform(
```
Comment (reedwm) [inline xla/service/gpu/cudnn_norm_rewriter.cc:592]: You add a lot of OptionalSupportedTransform calls. The PR description does state we now support "additional optional type conversions", but this also allows additional reshapes. The AdjustedDimensions function can take into account these reshapes, but I think it is not tested. Update the PR description to state additional reshapes removing/adding degenerate dimensions are allowed, and also add a test with such a reshape.

**Review by reedwm — CHANGES_REQUESTED** [review]

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 4]
```diff
@@ -171,6 +172,42 @@ bool CompatibleElementType(const HloInstruction* instr) {
   return element_type == BF16 || element_type == F16 || element_type == F32;
 }
 
+// Returns the dimensions associated with shape, adjusted for the removal of any
+// degenerate dimensions in shape.
```
Comment (reedwm) [inline xla/service/gpu/cudnn_norm_rewriter.cc:176]: Maybe add the sentences:

Specifically, for each dimension 'd' in dimensions, returns the new index of d if all dimensions of size 1 are removed from 'shape'. If d itself has size 1, it is not included in the returned vector.

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 5]
```diff
@@ -275,15 +309,12 @@ bool FindTarget(const HloInstruction* custom_call, const HloInstruction* instr,
...
-  // The original and reshaped shape must not have degenerate dimensions.
-  if (ShapeUtil::HasDegenerateDimensions(original_shape) ||
-      ShapeUtil::HasDegenerateDimensions(reshaped_shape)) {
-    return {};
-  }
-
   auto dimension_product =
       [](const Shape& shape,
          absl::Span<const int64_t> product_dimensions) -> int64_t {
+    if (product_dimensions.empty()) {
```
Comment (reedwm) [inline xla/service/gpu/cudnn_norm_rewriter.cc:315]: Why return 0 here instead of 1?

File: xla/service/gpu/cudnn_norm_rewriter.cc [hunk 6]
```diff
@@ -930,25 +969,30 @@ class CudnnNormRewriterVisitor : public DfsHloRewriteVisitor {
...
       // normalized are the leading dimensions.
-      std::vector<int64_t> non_norm_dims;
+      std::vector<int64_t> non_norm_dims, non_norm_dims_adjusted;
       for (int64_t x_dim = 0; x_dim < x.Instr()->shape().rank(); ++x_dim) {
         if (std::find(norm_dims.begin(), norm_dims.end(), x_dim) ==
             norm_dims.end()) {
           non_norm_dims.emplace_back(x_dim);
         }
       }
+      non_norm_dims_adjusted =
```
Comment (reedwm) [inline xla/service/gpu/cudnn_norm_rewriter.cc:993]: Better to not declare this above but instead here, and do: `std::vector<int64_t>non_norm_dims_adjusted = AdjustedDimensions(x.Instr()->shape(), non_norm_dims);`

**Review by reedwm — CHANGES_REQUESTED** [review]

File: xla/service/gpu/cudnn_norm_rewriter.cc (hunk 4 above)
Comment (SandSnip3r) [reply xla/service/gpu/cudnn_norm_rewriter.cc:176]: This amendment greatly improved my understanding of the function without needing to look at the actual implementation of the function.

**Review by reedwm — APPROVED** [review]
